### DISCUSSION TITLE
New in v11.0.0 - ✨References can be pinned in the chat buffer

### DISCUSSION THREAD

**Discussion by olimorris** [olimorris/codecompanion.nvim#575] [Announcements]
Hope you're all enjoying the holiday season. I've totally needed the downtime and have been able to crank out v11.0.0 in the last day which had been on the cards for sometime. Alas, life and other priorities take over.

## Introducing Pins

I'm pleased to announce that in the latest release you can now "pin" buffers and files in the chat buffer.

**What does this mean exactly?**

Well it means that every time you send a message to an LLM, the plugin will fetch the content from the pinned buffer/file and add it alongside your message. Previously, you would have had to manually re-add the contents of the reference to the message stack, which was a huge nuisance especially if you were using the `@editor` tool.

https://github.com/user-attachments/assets/e9fd9f85-f905-4c8a-b897-4bbd1f5e4afa

**How do I use it?**

In the chat buffer, move over a reference and press `gp` to pin it. You will see a pin icon appear next to the reference. This icon can also be changed in your config. Of course, you can also unpin a reference. As seen in the video above, I actually choose to pin the reference in the second message but could have done it in the first.

**What actually happens?**

The whole contents of the buffer or file is added to the message stack before your current message. This happens every time for the duration that the reference is pinned. So please be mindful of token consumption.

**Where does it work?**

Pins work for the _file_ and _buffer_ slash commands and also for the `#buffer` variable.

## Breaking Changes

There are two breaking changes in this release:

1. Agents have moved from `strategies.agent` to `strategies.chat.agents`. This should have happened looooong ago!
2. If you have custom variables that you use in the chat buffer, you'll need to ensure that the `execute` function within their class is renamed to `output`. You will also need to ensure that they implement a `self.Chat.add_message` function to add content to the chat buffer's message stack.

## Roadmap for `v11.*`

Things I'm aiming for in `v11.*`:

- Performance improvements - I am going to scope Tree-sitter queries for better improvement across large chat buffers as well as investigate #552 in more detail
- Allow the `@editor` agent to work with the _File_ and _Buffer_ slash commands. Currently it's scoped to work with line numbers and those functions just share the contents without line numbers
- Knowledge - I want to make it easier to open chat buffers with pre-defined references. This will make conversations more productive as you're not having to manually add knowledge (from the _Fetch_ or _Help_ slash commands)
- Automated Editing - I will strive to make it much easier to have the `@editor` and `@cmd_runner` tools working alongside pins to edit buffers or files, run tests, see the impact and determine next steps.

As always, if you like this release, I'm always open to a [coffee](https://buymeacoffee.com/olimorris) 😃.

Wishing you all the best for 2025!!

**Comment by LintaoAmons** [comment]
I found the `@editor` tool is not very intuitive to use. There's only one shortcut to accept all the changes by hit `ga` right?

Or I have not got the right way to use it?

**Comment by bassamsdata** [comment]
Thank you for all those updates. Congratulations on reaching v11 as well!

As I understood from your plans for @editor, currently, only the #buffer variable is sending line numbers to the LLM, is that correct?

I’ve always wondered why we have the /buffer slash command and the #buffer variable (even though I know I can send multiple buffers in slash commands). Personally, I prefer sending the line numbers directly to the LLM.

**Reply by bassamsdata** [reply]
One more question, if I may: Is the logic for the pins entirely within the chat/init.lua file? because I have an idea in mind, but it hasn’t fully taken shape yet.

**Reply by olimorris** [reply]
> I’ve always wondered why we have the /buffer slash command and the #buffer variable

Principally because the latter was implemented first. I don't want to disrupt someone's workflow by removing it and forcing them to adopt slash commands. Also, variables are just quicker to add to the chat buffer.

> As I understood from your plans for @editor, currently, only the #buffer variable is sending line numbers to the LLM, is that correct?

Yep. Because I'm so time constrained, I'll generally wait until someone shouts about something before changing/implementing it. I had forgotten that the `buffer` slash command didn't include line numbers and therefore the `@editor` tool didn't work, so was surprised to see no one shout about it. Can only deduce that people mostly use `#buffer` when working with the tool.

**Reply by olimorris** [reply]
> is the logic for the pins entirely within the chat/init.lua file? because I have an idea in mind, but it hasn’t fully taken shape yet.

...and the chat/keymaps.lua and chat/references.lua. Might also be worthwhile referring to the tests that accompanied the feature.

**Reply by bassamsdata** [reply]
Thank you for all this information.

> Can only deduce that people mostly use #buffer when working with the too

alright, that's me as well :)

> ...and the chat/keymaps.lua and chat/references.lua. Might also be worthwhile referring to the tests that accompanied the feature.

Ah, I always forget that tests might be useful. Thank you

**Comment by GitMurf** [comment]
@olimorris so the new pinning functionality is working wonderfully! Nice work!

I do have one question / "issues" in reference to your quoted statement from above:

> The whole contents of the buffer or file is added to the message stack before your current message. This happens every time for the duration that the reference is pinned. So please be mindful of token consumption.

You are indeed correct that the token count increases quite drastically depending on what file(s) you pin. One thing that would help tremendously though is if when you pin a buffer/file, you can have it remove the old messages that have the old copies of the buffer content. My main use case is to just want to be able to keep chatting with the most up to date buffer content (which the pin does) and rarely have use case of needing the countless other duplicate messages with the buffer content (at old states). The main problem is this exponentially increases the token count as each time you submit a message that is pinned, the token count just for the buffer contents turns into `Token Count in Buffer * Message Count`. A 500 line file has roughly 5,000 tokens in it, so after just 5 messages my token count for the buffers is 25,000 when in reality all I want is to keep submitting an updated copy of the same buffer content in a single message keeping a constant token count per buffer / file (5,000 tokens in this example).

It feels like maybe there is room for another "state" for pins where you have 1. Un-pinned 2. Pinned (as currently works) 3. Pinned (latest buffer contents only... removes the previous buffer content messages from the history).

Does this make sense? Thoughts?

**Reply by bassamsdata** [reply]
I’m not Oli, of course, but if I may chime in for a moment, I’m currently working on a feature that addresses this exact issue, so there’s no need to keep sending the buffer repeatedly. That’s why I was asking about where the logic for the pins is located.

It’s not quite finished yet, but I’m hoping to have a PR ready by the weekend, of course, assuming Oli is on board with the idea.

Apologies for jumping in! 😊

**Reply by GitMurf** [reply]
@bassamsdata sounds great! If you want a tester or any help or review, let me know!

**Reply by bassamsdata** [reply]
Great! Thanks a lot. I’ll mention you once I’ve made some progress.

**Reply by GitMurf** [reply]
@olimorris I am finding that the ability to do custom keymaps with access to the Chat instance in the callback is extremely powerful! ❤️

I am able to get access to chat messages and delete ones I want to pretty easily, but the remaining problem I am having is "syncing" the chat buffer content after deleting messages (i.e., remove the deleted messages from the chat buffer window and have it "refresh" the extmarks etc.). I have explored all the different methods I could think of after poking into the Chat / UI / References etc. methods, but I have not found the right method (or combo of methods) to do this. 

Is there a simple / straightforward way to do this? Essentially reload the chat buffer based off the current Chat.messages state.

I even tried clearing the chat buffer window and trying to loop through the current messages and add them back in one at a time but have not found success with anything I have tried.

Perfect world would be a method to "refresh/reload chat window buffer based off current state of Chat.messages and Chat.references etc.". This would seemingly also be a step towards being able to save chat conversations and reload them later to continue conversation 🤔 🙂

Let me know if you have any thoughts or advice. For now I am totally fine with something potentially "hacky" until there are more "official" primitives available (or maybe there already are but I just couldn't find them!).

**Reply by olimorris** [reply]
> One thing that would help tremendously .. you can have it remove the old messages that have the old copies of the buffer content.

> It feels like maybe there is room for another "state" for pins where you have 1. Un-pinned 2. Pinned (as currently works) 3. Pinned (latest buffer contents only... removes the previous buffer content messages from the history).

My initial concerns with pins were solely around token management. I thought about enabling them for the symbol slash command but ultimately thought that would be somewhat useless for most users. I also thought about clearing out previously pinned buffers as well. Ultimately, I left that out of Pins v1.0 because I expected it wouldn't play well for those of us who run local LLMs. Also, I strongly prefer not to force change the history of a conversation for fear of sub-par responses and the dreaded h word. My expectation was after using the feature, users would unpin a buffer for future responses then re-pin it if required, if they're aiming to minimize tokens.

All this being said, I think a third state _could_ be an option. Although, I'd want that ability to be scoped down to an adapter (possibly even model) level to allow local LLM users to turn it off. I'd need to think of a way to do that elegantly.

The events in the plugin (`CodeCompanionRequestFinished`) alongside `require("codecompanion").buf_get_chat(bufnr)` would allow you to manage messages and delete pinned references yourself. I may need to make it easier detect a pin from the messages table but that could be an option.

**Reply by GitMurf** [reply]
> The events in the plugin (CodeCompanionRequestFinished) alongside require("codecompanion").buf_get_chat(bufnr) would allow you to manage messages and delete pinned references yourself. I may need to make it easier detect a pin from the messages table but that could be an option.

See my most recent message above. To your point, I actually got what I needed for deleting messages by creating a custom keymap and using the provided Chat param to get messages and then able to get reference ID from chat message if has reference and then can get a reference from its ID by using Chat.References and then check if ref is pinned by looking at ref.opts.pinned == true... on my phone right now so this is off top of my head but mostly accurate ;)

As described above, my only issue I am still facing is figuring out how to "reload" / "refresh" the chat buffer to account for deleted messages (when deleting other messages because references are typically set to visible false so no need to refresh chat buffer... but for deleting a bad partial message for example when Claude API craps out in middle as an example) to remove them from chat buffer (and make sure all the extmarks are updated accordingly).

**Reply by olimorris** [reply]
I've just added `CodeCompanionChatPin` as an event. Every time a pinned reference is refreshed, this fires and includes the chat buffer number alongside the ID of the reference.

**Reply by bassamsdata** [reply]
> @bassamsdata sounds great! If you want a tester or any help or review, let me know!

Hey @GitMurf, I’ve been working on PR #610 for the past week and it’s now in the final stages. I’d like to know if you’d be able to test it out. It’s not completely finished yet, but I’m trying to figure out the best style to send it to the LLM.

**Comment by taketwo** [comment]
Hi, I've just discovered this amazing plugin and I'm reading through the documentation website and announcements history to learn about the existing features. I must admit, I'm confused about the file/buffer reference sharing and pinning.

From the documentation on References:

> Sharing context with an LLM is a crucial in order to generate useful responses. In the plugin, references are defined as output that is shared with a chat buffer via a Variable, Slash Command or Agent/Tool. They appear in a blockquote entitled Sharing. In essence, this is context that you're sharing with an LLM.

To me, this description sounds as if the shared file would be re-added to each next message. This blockquote with _present continuous_ tense "sharing" reinforces that impression:

![image](https://github.com/user-attachments/assets/8b126bf0-30c8-4606-a781-09d8ff0465e3)

Basically, the mental model I have when looking at the chat window is that this area highlighted by red rectangle is the visual representation of the message that I'm crafting. It will consist of my prompt text and of listed files that I'm sharing.

Plus, the following passage from the documentation:

> References can be removed from the chat buffer by simply deleting them. On the next turn, the context will be removed from the message history.

gives me the idea that unless I delete the lines with the references, it will be sent to the LLM.

In summary, reference sharing already sounds like it should behave as described by this new pinning feature. But if this was the case, you wouldn't introduce pinning, I reckon :)  So... how do unpinned/unwatched references work? What does it mean if such reference is "removed from the message history"?

**Reply by olimorris** [reply]
Appreciate the feedback.

So references are shared once with the LLM (I will make that explicit in the README), but remain for the duration of your conversation. The sharing section reminds you that is context the LLM currently has. Pins and Watchers send over an updated copy of the reference. The latter does so more intelligently for buffers.

**Reply by taketwo** [reply]
Thanks for the answer! It's still not entirely clear to me what happens when I delete the blockquote from the message I'm crafting. According to the documentation, "the context will be removed from the message history". But does that mean in practice? Assuming the reference was neither pinned nor watched, it was not going to be re-added to the conversation anyway, so what difference would deleting these lines make?

Based on your answer, I understand that the meaning of the blockquote is somewhat different in the initial and subsequent messages:
* In the initial message, when you've just added a reference, the meaning is "These references will be shared with LLM as part of this message"
* In subsequent messages, the meaning is "These references have been shared with LLM as part of one of the previous messages"

Is that right?

**Reply by olimorris** [reply]
I'm not at my computer tonight so I'll save a full reply for tomorrow. 

The quickest way is to play for yourself. `gd` reveals the full message stack where you can see what's happening under the hood wrt to all things referenced. Pin, watch and delete and you'll see how they change.

**Reply by olimorris** [reply]
Hopefully it's clearer now: https://codecompanion.olimorris.dev/usage/chat-buffer/#references

**Reply by taketwo** [reply]
Thanks for taking time to answer and update the documentation. The `gd` command is very useful! I can now see how deleting a shared reference literally removes the corresponding message from the history.

It was still hard to wrap my head around the implication of modifying the history this way. As in, I'm familiar with editing an older message in ChatGPT or Claude chat, but that creates a branch in the conversation. Here, in CodeCompanion, we just continue like nothing happened and this is confusing. I ran the following experiment to understand the behavior:

![image](https://github.com/user-attachments/assets/a92ea44a-e690-454b-8780-2dd1c11407f3)

In the top instance of Neovim, I keep the reference and LLM can answer my questions about the file contents correctly. In the bottom Neovim instance, I remove the reference before sending the second message. The file contents are dropped from the chat history and the model can no longer answer the question correctly.

Now that I understand this behavior, it makes we wonder: instead of deleting a reference in the history, can we modify its content? Why didn't you implement the pinning feature this way?

**Reply by taketwo** [reply]
I found your answer in another thread:

> I strongly prefer not to force change the history of a conversation for fear of sub-par responses and the dreaded h word.

That explains why pinning is done the way it is done. So what remains for me to understand, is why did you decide to allow chat history alternation through reference deletion? As you noted, this may lead to sub-par responses. From my side, I can add that it's really confusing that by crafting a new message we have the power to alternate the previous chat history.

**Reply by olimorris** [reply]
I found, on occasion, when adding multiple files via Telescope, I may accidentally select one I didn't intend to. I delete it before the LLM ever sees it, thus not interfering with the history. That was my original intention. It's implementation does remove all references but ideally a user doesn't get n responses in (where n > 1) and start removing references.

**Reply by olimorris** [reply]
> Why didn't you implement the pinning feature this way?

Crikey, tough interview 😂. I can't recall my thought process for that particular feature.

A little further down the line I am planning for the message history to be editable but I need to firm up how I want that to feel from a user perspective.

**Reply by taketwo** [reply]
Got it! Sorry for giving you the third degree about your designs :see_no_evil: Anyway, the most important thing I want to say is: thanks for all the hard work writing and maintaining this plugin!

**Reply by GitMurf** [reply]
@taketwo the best way to think about this is that every message you send is really a "fresh new conversation". What I mean by that is, there really is no such thing as "history" with LLMs. Each message you send you attach the entire conversation and the LLM essentially "re-creates that conversation" each time based off what you send it.

So if you are about to send message 5 in the conversation and delete a reference (file attached in message 2 lets say) the mental model is that you are sending this conversation without message 2 as if it never happened.

It is a mental model that took me a while to fully grasp and I think a lot of people "forget" / conflate what a "conversation" with an LLM really means. Really it is not a conversation but a single message each time with other prior context (which we consider a "conversation" as a whole).

**Reply by taketwo** [reply]
Thanks for the explanation, that makes sense.

> So if you are about to send message 5 in the conversation and delete a reference (file attached in message 2 lets say) the mental model is that you are sending this conversation without message 2 as if it never happened.

I guess what made it hard for me to understand this is that the plugin UI does not support that mental model. An attached reference is nothing else but a single message in the history. But the plugin displays it as if it was a recurring part of each next message.

**Reply by GitMurf** [reply]
> An attached reference is nothing else but a single message in the history. But the plugin displays it as if it was a recurring part of each next message.

Yeah it is hard to visualize the idea of a LLM conversation while also having it make sense from an end user presentation. I agree it isn't perfect but for me personally it is better than most tools IMO.

The last remaining thing I'd love is be able to delete messages just like you can references and have the UI reflect that as well.
